gamemodes/terrortown/entities/entities/ttt_basegrenade_proj.lua

```diff
@@ -0,0 +1,170 @@
+---
+-- common grenade projectile code
+-- @class ENT
+-- @section ttt_basegrenade_proj
+
+AddCSLuaFile()
+
+ENT.Type = "anim"
+ENT.Model = Model("models/weapons/w_eq_flashbang_thrown.mdl")
+ENT.playerDropWeaponOnContact = false
+ENT.sphericalPhysicsRadius = 0
+ENT.damageTriggerProjectileExplosion = true
+ENT.preventDamageChainReaction = true
+ENT.bounceFactor = 0
+ENT.bounceMomentumLoss = 0
+ENT.impactSoundSpeed = 50
+ENT.throwSound = Sound("entities/entities/ttt_basegrenade_proj/toss.wav")
+ENT.impactSound = Sound("entities/entities/ttt_basegrenade_proj/impact.wav")
+
+---
+-- @realm shared
+function ENT:SetupDataTables()
+	self:NetworkVar("Float", 0, "ExplodeTime")
+	self:NetworkVar("Bool", 0, "ShotToActivate")
+	self:NetworkVar("Bool", 1, "FailedToActivate")
+	self:NetworkVar("Bool", 2, "Exploded")
+	self:NetworkVar("String", 0, "LaunchWeaponClass")
+	self:NetworkVar("Entity", 0, "Thrower")
+
+	if SERVER then
+		self:SetExplodeTime(0)
+		self:SetFailedToActivate(false)
+		self:SetExploded(false)
+		self:SetShotToActivate(false)
+	end
+end
+
+---
+-- @realm shared
+function ENT:Initialize()
+	self:AddEFlags( EFL_FORCE_CHECK_TRANSMIT )
+	self:SetModel(self.Model)
+
+	if self.sphericalPhysicsRadius > 0 then
+		self:PhysicsInitSphere(self.sphericalPhysicsRadius, "metal")
+		self:SetMoveType( MOVETYPE_VPHYSICS )
+		self:SetSolid(SOLID_VPHYSICS)
+	else
+		self:PhysicsInit(SOLID_VPHYSICS)
+		self:SetMoveType(MOVETYPE_VPHYSICS)
+		self:SetSolid(SOLID_BBOX)
+	end
+	self:SetCollisionGroup(COLLISION_GROUP_NONE)
+end
+
+---
+-- Everything that needs cleaning up prior to removal, but OnRemove is too late for.
+-- @realm shared
+function ENT:Destroy()
+	if CLIENT then return end
+	self:SetExploded(true)
+	self:SetDetonateExact(0)
+	self:AddEFlags(EFL_FORCE_CHECK_TRANSMIT)
+	self:AddEFlags(EFL_DORMANT)
+	if SERVER then self:Remove() end
+end
+
+---
+-- @param number length How far away _from now_ to detonate.
+-- @realm shared
+function ENT:SetDetonateTimer(length)
+	self:SetDetonateExact( CurTime() + length )
+end
+
+---
+-- @param nil|number t The time we should explode, or <code>CurTime()</code> if falsy.
+-- @realm shared
+function ENT:SetDetonateExact(t)
+	self:SetExplodeTime(t or CurTime())
+end
+
+---
+-- override to describe what happens when the nade explodes
+-- @param TraceResult tr
+-- @realm server
+function ENT:Explode(tr)
+	ErrorNoHalt("ERROR: BaseGrenadeProjectile explosion code not overridden!\n")
+end
+
+---
+-- Attempt to detonate the projectile immediately.
+-- This was externalized to permit detonation outside of the throw / release logic. (Mid-Air, other phenomenon)
+-- @realm shared
+function ENT:TryExplode()
```

Comment:
does this need to be run on the client at all? The Client does nothing here besides setting some unused locals, right?

Reply:
Previously, the client would be the one to create decals, but since decals have been made server authoritative in the time between this code being written originally and now, it only appears that it isn't that way. Some old grenades that inherit from basegrenade for TTT1 might still depend on this behavior.

Reply:
in that case we should probably guard against the client entering this function to far.

If we guard against the client at the start of the function, the if-clauses later don't need the SERVER as well.

Reply:
If we guard against this function then we guard against Explode being called, with old grenades still having CLIENT specific code in them. I think if we make it unshared then it will break existing grenades that inherit from this.

In general I'd prefer to minimize the amount of parts of the code where something is shared but has an undocumented lack of effect on the client/server. The existing gmod api is already full of those situations where I can't tell at a glance if a function is exposed just by coincidence but still a no-op. For existing APIs, those are kind of the breaks, and this is one where it is merely extending an existing API.